### PR TITLE
Add GeoportailFrance tilelayers to providers.json.

### PR DESCRIPTION
**Pull request by HaudinFlorence**
This PR proposes a new version of the json providers for the basemaps, including more than 250 additionnal tilelayers from Geoportail France. It also includes a piece of python code added to provider_sources/_compress_providers.py, used to generate the new providers from the xml requests. The names of the tilelayers have been automated from the original names and may be revisited if needed to reduce length/complexity.

### CODE REVIEW

**Comment by martinfleis** [comment]
Hi, thanks a lot for this massive expansion of the database. 

To make it work within our current setup, the code that is currently in your `generate_providers.py` file should probably become a part of `provider_sources/compress_providers.py`. That way we always load fresh set from GeoportailFrance anytime we refresh the database. That is an automatised process composed of three steps right now - 1. fetch data from leaflet-providers, 2. merge them with additions specified in `xyzservices-providers.json` and 3. compress the result and save it as `data/providers.json`. I guess that if we add your code after the second step, the whole pipeline should work without any other changes. Apart from updating `update_providers.yaml` to include `xmltodict` and `requests` to make it run on GHA.

**Comment by HaudinFlorence** [comment]
> Hi, thanks a lot for this massive expansion of the database.
> 
> To make it work within our current setup, the code that is currently in your `generate_providers.py` file should probably become a part of `provider_sources/compress_providers.py`. That way we always load fresh set from GeoportailFrance anytime we refresh the database. That is an automatised process composed of three steps right now - 1. fetch data from leaflet-providers, 2. merge them with additions specified in `xyzservices-providers.json` and 3. compress the result and save it as `data/providers.json`. I guess that if we add your code after the second step, the whole pipeline should work without any other changes. Apart from updating `update_providers.yaml` to include `xmltodict` and `requests` to make it run on GHA.

Hello. Thanks for the review. The suggestions proposed should have been taken into account in commit b9f4f5f.

**Comment by HaudinFlorence** [comment]
please update galata reference

**Review by martinfleis — COMMENTED** [review]
Thanks! It looks great. I just compressed the JSON to include these in tests and committed that to check if all of these work as expected.

File: doc/source/_static/generate_gallery.js [hunk 1]
```diff
@@ -260,9 +265,9 @@ function initLeafletGallery(el) {
                     newData = val;
 
                     for ([newKey, newVal] of Object.entries(newData)) {
-                        if (newVal["bounds"] !== undefined) {
+                        /*if (newVal["bounds"] !== undefined) {
```
[the remark this answers is not in the document]
Comment (HaudinFlorence) [reply doc/source/_static/generate_gallery.js:268]: The condition `if (newVal["bounds"] !== undefined) {` can be commented or removed.

**Comment by HaudinFlorence** [comment]
> Thanks! It looks great. I just compressed the JSON to include these in tests and committed that to check if all of these work as expected.

Ok. Thanks. Hoping everything will work correctly !

**Comment by martinfleis** [comment]
hmm, quite a few of these providers failed on CI (see https://github.com/geopandas/xyzservices/runs/8248560664?check_suite_focus=true). Can you have a look why is that? One reason may be that we use centre of bounds in the tile query https://github.com/geopandas/xyzservices/blob/42e65c9af8434bb01c177d3eade8ad6d21fb5947/xyzservices/tests/test_providers.py#L19-L28

If some of these cover France and its overseas territories, I assume there is nothing in the centre. It is not the provider that is affected and we have this list of tiles where I always add one I know should be present whenever this occurs. Can you try to debug this and possibly add some known tiles to make CI happy?

https://github.com/geopandas/xyzservices/blob/42e65c9af8434bb01c177d3eade8ad6d21fb5947/xyzservices/tests/test_providers.py#L63-L71

**Comment by HaudinFlorence** [comment]
> hmm, quite a few of these providers failed on CI (see https://github.com/geopandas/xyzservices/runs/8248560664?check_suite_focus=true). Can you have a look why is that? One reason may be that we use centre of bounds in the tile query
> 
> https://github.com/geopandas/xyzservices/blob/42e65c9af8434bb01c177d3eade8ad6d21fb5947/xyzservices/tests/test_providers.py#L19-L28
> 
> If some of these cover France and its overseas territories, I assume there is nothing in the centre. It is not the provider that is affected and we have this list of tiles where I always add one I know should be present whenever this occurs. Can you try to debug this and possibly add some known tiles to make CI happy?
> 
> https://github.com/geopandas/xyzservices/blob/42e65c9af8434bb01c177d3eade8ad6d21fb5947/xyzservices/tests/test_providers.py#L63-L71

I can have a look. I am not sure how to proceed (I am quite newbie with fixing CI test).

**Comment by HaudinFlorence** [comment]
I tried to add 2 tiles that are (if not mistaken) somewhere in the middle of France and another one in Paris. But it doesn't seem to solve the failing ci test issue.

**Comment by martinfleis** [comment]
I'll have a look either later today or in coming days.

**Comment by martinfleis** [comment]
We're down to 18 failures from initial 37! Good!

**Comment by HaudinFlorence** [comment]
> We're down to 18 failures from initial 37! Good!

Yes. I am trying to reduce progressively the numbers of failures. Sorry for the multiple pushes.

**Comment by HaudinFlorence** [comment]
> We're down to 18 failures from initial 37! Good!

I spent some time to try to get rid off the remaining failures but I couldn't, even when adding known tiles for which I can see something when displaying the tile layer with Leaflet. So I propose for the moment to exclude these 18 tile layers when producing the providers.json with `_compress_providers.py`? What do you think about it ? Or if you have any other idea, please let me know.

**Comment by martinfleis** [comment]
Let me have a look at it later this week/early next week.

**Comment by HaudinFlorence** [comment]
OK. I may miss something about the tests and adding tiles in the list of tiles known to work. Because here is an example with `Geoportail.Securoute_Te_All`, that is part of the failing tests. I can see some something on the map (some dots ) when displaying the tile layer with leaflet in a js script, for instance on tile  (9, 181, 259) that is part of the map.
Screenshot from 2022-09-09 18-12-13

But I have tried to add this tile in the list of known ones to test and it did not fix the failing test.

**Comment by martinfleis** [comment]
@HaudinFlorence I've added two other tiles and got it down to 9 layers returning 404. Not sure how to make those work, don't see them in leaflet either. So I marked them as `broken`. They are still included but our CI now doesn't fail when these don't resolve.

**Comment by HaudinFlorence** [comment]
OK. Thanks for the feedback and for reducing the number of failing layers.

**Comment by martinfleis** [comment]
Thanks a lot for this massive contribution ;).

**Pull request merged**